src/unicode/fold.c

```diff
@@ -0,0 +1,135 @@
+    BSON_ASSERT_PARAM(str);
+    BSON_ASSERT_PARAM(out_str);
+    BSON_ASSERT_PARAM(out_len);
+    BSON_ASSERT_PARAM(status);
+
+    if (!(options & (kUnicodeFoldRemoveDiacritics | kUnicodeFoldToLower))) {
+        CLIENT_ERR("unicode_fold: Either case or diacritic folding must be enabled");
+        return false;
+    }
+    *out_str = bson_malloc(len);
+    const char *input_it = str;
+    const char *end_it = str + len;
+    char *output_it = *out_str;
+    while (input_it < end_it) {
+        const uint8_t first_byte = (uint8_t)*input_it++;
+        bson_unichar_t codepoint = 0;
+        if (first_byte <= 0x7f) {
+            // ASCII special case. Can use faster operations.
+            if ((options & kUnicodeFoldToLower) && (first_byte >= 'A' && first_byte <= 'Z')) {
+                codepoint = first_byte | 0x20; // Set the ascii lowercase bit on the character.
+            } else {
+                // ASCII has two pure diacritics that should be skipped, and no characters that
+                // change when removing diacritics.
+                if ((options & kUnicodeFoldRemoveDiacritics) && (first_byte == '^' || first_byte == '`')) {
+                    continue;
+                }
+                codepoint = first_byte;
+            }
+        } else {
+            // Multi-byte character
+            size_t leading_ones = mc_count_leading_zeros_u64(~(((uint64_t)first_byte) << (64 - 8)));
+
+            // Only checking enough to ensure that this code doesn't crash in the face of malformed
+            // utf-8. We make no guarantees about what results will be returned in this case.
+            if (!(leading_ones > 1 && leading_ones <= 4 && input_it + (leading_ones - 1) <= end_it)) {
+                CLIENT_ERR("unicode_fold: Text contains invalid UTF-8");
+                free(*out_str);
+                return false;
+            }
+
+            codepoint = (bson_unichar_t)(first_byte & (0xff >> leading_ones)); // mask off the size indicator.
+            for (size_t sub_byte_index = 1; sub_byte_index < leading_ones; sub_byte_index++) {
+                const uint8_t sub_byte = (uint8_t)*input_it++;
+                codepoint <<= 6;
+                codepoint |= sub_byte & 0x3f; // mask off continuation bits.
+            }
+
+            if (options & kUnicodeFoldToLower) {
+                bson_unichar_t new_cp = unicode_codepoint_to_lower(codepoint);
+                codepoint = new_cp;
+            }
+
+            if ((options & kUnicodeFoldRemoveDiacritics)) {
+                codepoint = unicode_codepoint_remove_diacritics(codepoint);
+                if (!codepoint) {
+                    continue; // codepoint is a pure diacritic.
+                }
+            }
+        }
+
+        if (!append_utf8_codepoint(codepoint, &output_it, status)) {
+            free(*out_str);
+            return false;
+        }
+    }
+
+    *out_len = (size_t)(output_it - *out_str);
+    *out_str = realloc(*out_str, *out_len);
```

Comment:
I personally wouldn't bother with the `realloc` here just to shrink it. The folded string is not gonna be around long enough to make it worth the realloc cost.

Need to null terminate output string:
```suggestion
    *output_it = '\0';
    *out_len = (size_t)(output_it - *out_str);
```